controllers/mysql_controller.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aiven/aiven-go-client"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aiven/aiven-operator/api/v1alpha1"
+)
+
+// MySQLReconciler reconciles a MySQL object
+type MySQLReconciler struct {
+	Controller
+}
+
+//+kubebuilder:rbac:groups=aiven.io,resources=mysqls,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=aiven.io,resources=mysqls/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=aiven.io,resources=mysqls/finalizers,verbs=update
+
+func (r *MySQLReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return r.reconcileInstance(ctx, req, newGenericServiceHandler(newMySQLAdapter), &v1alpha1.MySQL{})
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MySQLReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.MySQL{}).
+		Owns(&corev1.Secret{}).
+		Complete(r)
+}
+
+func newMySQLAdapter(object client.Object) (serviceAdapter, error) {
+	mysql, ok := object.(*v1alpha1.MySQL)
+	if !ok {
+		return nil, fmt.Errorf("object is not of type v1alpha1.MySQL")
+	}
+	return &mySQLAdapter{mysql}, nil
+}
+
+// mySQLAdapter handles an Aiven MySQL service
+type mySQLAdapter struct {
```

Comment:
Something for the future, perhaps all the adapter functions accept `newSecret` can be generalized further to avoid a copy&paste when we expand this logic to other services.